### ISSUE TITLE
Got a unknown ext type error but didn't use forward plugin

### ISSUE THREAD

**Issue by zhigang** [fluent/fluent-plugin-kafka#243]
I got a lot of error when send message to kafka. I didn't use forward plugin.
My environment is in kubernetes cluster.
Please help. Thanks.

* Version

  * kafka:1.1.0
  * fluentd: 1.3.3
  * fluent-plugin-kafka: 0.8.4
  * ruby-kafka: 0.7.5

* Log

```
2019-02-15 06:34:42 +0000 [warn]: unexpected error during format record. Skip broken event: error="can't dup NilClass" error_class="TypeError" time=31 record=nil
2019-02-15 06:34:42 +0000 [warn]: unexpected error during format record. Skip broken event: error="can't dup NilClass" error_class="TypeError" time=31 record=nil
2019-02-15 06:34:42 +0000 [warn]: Send exception occurred: unexpected extension type
2019-02-15 06:34:42 +0000 [warn]: Exception Backtrace : /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `each'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `block in each'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/buffer/file_chunk.rb:172:in `open'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:322:in `each'
/var/lib/gems/2.3.0/gems/fluent-plugin-kafka-0.8.4/lib/fluent/plugin/out_kafka2.rb:204:in `write'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1123:in `try_flush'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1423:in `flush_thread_run'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:452:in `block (2 levels) in start'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-02-15 06:34:42 +0000 [warn]: Send exception occurred: unexpected extension type
2019-02-15 06:34:42 +0000 [warn]: Exception Backtrace : /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `each'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `block in each'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/buffer/file_chunk.rb:172:in `open'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:322:in `each'
/var/lib/gems/2.3.0/gems/fluent-plugin-kafka-0.8.4/lib/fluent/plugin/out_kafka2.rb:204:in `write'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1123:in `try_flush'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1423:in `flush_thread_run'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:452:in `block (2 levels) in start'
/var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-02-15 06:34:42 +0000 [warn]: failed to flush the buffer. retry_time=0 next_retry_seconds=2019-02-15 06:34:43 +0000 chunk="57f7b79651c6c75703ba35f3033aed77" error_class=MessagePack::UnknownExtTypeError error="unexpected extension type"
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `each'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:323:in `block in each'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/buffer/file_chunk.rb:172:in `open'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/event.rb:322:in `each'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluent-plugin-kafka-0.8.4/lib/fluent/plugin/out_kafka2.rb:204:in `write'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1123:in `try_flush'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:1423:in `flush_thread_run'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin/output.rb:452:in `block (2 levels) in start'
  2019-02-15 06:34:42 +0000 [warn]: /var/lib/gems/2.3.0/gems/fluentd-1.3.3/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
```

**Comment by repeatedly** [comment]
Could you write reproducible step with actual configuration?
Hard to debug with this information.

**Comment by zhigang** [comment]
@repeatedly This is my config file 
```
  system.conf: |-
    <system>
      root_dir /tmp/fluentd-buffers/
    </system>

  containers.input.conf: |-
    # This configuration file for Fluentd / td-agent is used
    # to watch changes to Docker log files. The kubelet creates symlinks that
    # capture the pod name, namespace, container name & Docker container ID
    # to the docker logs for pods in the /var/log/containers directory on the host.
    # If running this fluentd configuration in a Docker container, the /var/log
    # directory should be mounted in the container.
    #
    # These logs are then submitted to Elasticsearch which assumes the
    # installation of the fluent-plugin-elasticsearch & the
    # fluent-plugin-kubernetes_metadata_filter plugins.
    # See https://github.com/uken/fluent-plugin-elasticsearch &
    # https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter for
    # more information about the plugins.
    #
    # Example
    # =======
    # A line in the Docker log file might look like this JSON:
    #
    # {"log":"2014/09/25 21:15:03 Got request with path wombat\n",
    #  "stream":"stderr",
    #   "time":"2014-09-25T21:15:03.499185026Z"}
    #
    # The time_format specification below makes sure we properly
    # parse the time format produced by Docker. This will be
    # submitted to Elasticsearch and should appear like:
    # $ curl 'http://elasticsearch-logging:9200/_search?pretty'
    # ...
    # {
    #      "_index" : "logstash-2014.09.25",
    #      "_type" : "fluentd",
    #      "_id" : "VBrbor2QTuGpsQyTCdfzqA",
    #      "_score" : 1.0,
    #      "_source":{"log":"2014/09/25 22:45:50 Got request with path wombat\n",
    #                 "stream":"stderr","tag":"docker.container.all",
    #                 "@timestamp":"2014-09-25T22:45:50+00:00"}
    #    },
    # ...
    #
    # The Kubernetes fluentd plugin is used to write the Kubernetes metadata to the log
    # record & add labels to the log record if properly configured. This enables users
    # to filter & search logs on any metadata.
    # For example a Docker container's logs might be in the directory:
    #
    #  /var/lib/docker/containers/997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b
    #
    # and in the file:
    #
    #  997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b-json.log
    #
    # where 997599971ee6... is the Docker ID of the running container.
    # The Kubernetes kubelet makes a symbolic link to this file on the host machine
    # in the /var/log/containers directory which includes the pod name and the Kubernetes
    # container name:
    #
    #    synthetic-logger-0.25lps-pod_default_synth-lgr-997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b.log
    #    ->
    #    /var/lib/docker/containers/997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b/997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b-json.log
    #
    # The /var/log directory on the host is mapped to the /var/log directory in the container
    # running this instance of Fluentd and we end up collecting the file:
    #
    #   /var/log/containers/synthetic-logger-0.25lps-pod_default_synth-lgr-997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b.log
    #
    # This results in the tag:
    #
    #  var.log.containers.synthetic-logger-0.25lps-pod_default_synth-lgr-997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b.log
    #
    # The Kubernetes fluentd plugin is used to extract the namespace, pod name & container name
    # which are added to the log message as a kubernetes field object & the Docker container ID
    # is also added under the docker field object.
    # The final tag is:
    #
    #   kubernetes.var.log.containers.synthetic-logger-0.25lps-pod_default_synth-lgr-997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b.log
    #
    # And the final log record look like:
    #
    # {
    #   "log":"2014/09/25 21:15:03 Got request with path wombat\n",
    #   "stream":"stderr",
    #   "time":"2014-09-25T21:15:03.499185026Z",
    #   "kubernetes": {
    #     "namespace": "default",
    #     "pod_name": "synthetic-logger-0.25lps-pod",
    #     "container_name": "synth-lgr"
    #   },
    #   "docker": {
    #     "container_id": "997599971ee6366d4a5920d25b79286ad45ff37a74494f262e3bc98d909d0a7b"
    #   }
    # }
    #
    # This makes it easier for users to search for logs by pod name or by
    # the name of the Kubernetes container regardless of how many times the
    # Kubernetes pod has been restarted (resulting in a several Docker container IDs).

    # Json Log Example:
    # {"log":"[info:2016-02-16T16:04:05.930-08:00] Some log text here\n","stream":"stdout","time":"2016-02-17T00:04:05.931087621Z"}
    # CRI Log Example:
    # 2016-02-17T00:04:05.931087621Z stdout F [info:2016-02-16T16:04:05.930-08:00] Some log text here
    <source>
      @id fluentd-containers.log
      @type tail
      path /var/log/containers/*.log
      pos_file /var/log/es-containers.log.pos
      tag raw.kubernetes.*
      read_from_head true
      <parse>
        @type multi_format
        <pattern>
          format json
          time_key time
          time_format %Y-%m-%dT%H:%M:%S.%NZ
        </pattern>
        <pattern>
          format /^(?<time>.+) (?<stream>stdout|stderr) [^ ]* (?<log>.*)$/
          time_format %Y-%m-%dT%H:%M:%S.%N%:z
        </pattern>
      </parse>
    </source>

    # Detect exceptions in the log output and forward them as one log entry.
    <match raw.kubernetes.**>
      @id raw.kubernetes
      @type detect_exceptions
      remove_tag_prefix raw
      message log
      stream stream
      multiline_flush_interval 5
      max_bytes 500000
      max_lines 1000
    </match>
  
  monitoring.conf: |-
    # Prometheus Exporter Plugin
    # input plugin that exports metrics
    <source>
      @type prometheus
    </source>

    <source>
      @type monitor_agent
    </source>

    # input plugin that collects metrics from MonitorAgent
    <source>
      @type prometheus_monitor
      <labels>
        host ${hostname}
      </labels>
    </source>

    # input plugin that collects metrics for output plugin
    <source>
      @type prometheus_output_monitor
      <labels>
        host ${hostname}
      </labels>
    </source>

    # input plugin that collects metrics for in_tail plugin
    <source>
      @type prometheus_tail_monitor
      <labels>
        host ${hostname}
      </labels>
    </source>

  output.conf: |-
    # Enriches records with Kubernetes metadata
    <filter kubernetes.**>
      @type kubernetes_metadata
    </filter>

    <match **>
      @type kafka2
      brokers kafka-0.kafka:9092,kafka-1.kafka:9092,kafka-2.kafka:9092
      default_topic k8slog
      # default_message_key message
      # ruby-kafka producer options
      compression_codec gzip
      <inject>
        tag_key   tag
        time_key  time
        utc true
      </inject>
      <format>
        @type json
      </format>
      <buffer>
        @type file
        path /var/log/fluentd-buffers/kubernetes.system.buffer
        flush_mode interval
        flush_thread_count 2
        flush_interval 180s
        retry_type exponential_backoff
        retry_forever false
        retry_max_times 30
        retry_max_interval 120
        chunk_limit_size 32M
        queued_chunks_limit_size 128
        overflow_action block
        compress gzip
      </buffer>
    </match>
```

**Comment by zhigang** [comment]
@repeatedly I try to find reproducible step for this. but it doesn't work. I don't know what happened. But I need fix this as soon as possible. So I cleared the buffer. It running well for now. Anyway thanks for reply. ^.^

**Comment by coder-hugo** [comment]
I encountered the same issue with the elasicsearch output. I guess it's somehow related to the `compress gzip` configuration within the buffer section. I've configured this as well for the elasticsearch output. Every time I restart fluentd when there is still some data within the buffer I get the `unexpected extension type` error. If I disable the gzip compression for the buffer then everything is fine after restarting fluentd.
